File: arrays/10_unique_paths_with_all_cells/unique_paths_with_all_cells.py
import numpy as np


class Solution:
    def unique_paths_III(self, grid: np.ndarray) -> int:
        if not self._check_constraints(grid):
            raise ValueError('Constraints violated')
        non_obstacle = 0
        row, col = grid.shape
        for i in range(row):
            for j in range(col):
                if grid[i, j] >= 0:
                    non_obstacle += 1
                if grid[i, j] == 1:
                    start_x, start_y = i, j

        return self._backtrack(grid, start_x, start_y, non_obstacle)

    def _backtrack(self, grid: np.ndarray, x: int, y: int, remain: int) -> int:
        if (x < 0 or x >= grid.shape[0]) or (y < 0 or y >= grid.shape[1]) or grid[x, y] == -1:
            return 0
        if grid[x, y] == 2:
            return 1 if remain == 1 else 0

        # Mark the square as visited -> unreachable just like obstacle
        grid[x, y] = -1
        remain -= 1

        paths = self._backtrack(grid, x - 1, y, remain) + self._backtrack(grid, x + 1, y, remain) + self._backtrack(grid, x, y + 1, remain) + self._backtrack(grid, x, y - 1, remain)

        # Mark the square as unvisited to be considered by other possible paths
        grid[x, y] = 0
        remain += 1

        return paths

    def _check_constraints(self, grid: np.ndarray) -> bool:
        m = grid.shape[0]
        if m == 0 or m > 20:
            return False

        for i in range(m):
            n = grid[i].shape[0]
            if n == 0 or n > 20:
                return False
            if m * n == 0 or m * n > 20:
                return False
            for j in range(n):
                if grid[i, j] < -1 or grid[i, j] > 2:
                    return False

        return True

File: arrays/10_unique_paths_with_all_cells/unique_paths_with_all_cells.cpp
#include <iostream>
#include <stdexcept>
#include <vector>

using namespace std;


class Solution {
public:
    int uniquePathsIII(vector<vector<int>>& grid) {
        if (!checkConstraints(grid)) {
            throw std::runtime_error("Constraints violated");
        }
        int zeroCount = 0, startX = 0, startY = 0;
        for (int i = 0; i < grid.size(); i++) {
            for (int j = 0; j < grid[0].size(); j++) {
                if (grid[i][j] >= 0) zeroCount++;
                if (grid[i][j] == 1) {
                    startX = i;
                    startY = j;
                }
            }
        }
        return backtrack(grid, startX, startY, zeroCount);
    }

private:
    int backtrack(vector<vector<int>>& grid, int x, int y, int remain) {
        if (x < 0 || x >= grid.size() || y < 0 || y >= grid[0].size() || grid[x][y] == -1) {
            return 0;
        }
        if (grid[x][y] == 2) {
            return remain == 1 ? 1 : 0;
        }
        // Mark the square as visited -> unreachable just like obstacle
        grid[x][y] = -1;
        remain--;

        int totalPaths = backtrack(grid, x + 1, y, remain) +
                         backtrack(grid, x - 1, y, remain) +
                         backtrack(grid, x, y + 1, remain) +
                         backtrack(grid, x, y - 1, remain);

        // Mark the square as unvisited to be considered by other possible paths
        grid[x][y] = 0;
        remain++;

        return totalPaths;
    }

    // Helper function to check the constraints
    bool checkConstraints(const vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();
        if (m == 0 || m > 20) return false;
        if (n == 0 || n > 20) return false;

        for (auto& row : grid) {
            for (auto& cell : row) {
                if (cell < -1 || cell > 2) {
                    return false;
                }
            }
        }
        return true;
    }
};
